### PR TITLE
Rename rpc ops

### PR DESCRIPTION
**Pull request by typhoonzero**
This is part job of https://github.com/PaddlePaddle/Paddle/issues/6508

After this PR we can let users define very extensible distributed program using python.

### CODE REVIEW

File: paddle/operators/CMakeLists.txt [hunk 1]
```diff
@@ -122,9 +122,11 @@ if(WITH_DISTRIBUTE)
     set_source_files_properties(send_op.cc PROPERTIES COMPILE_FLAGS ${DISTRIBUTE_COMPILE_FLAGS})
     op_library(recv_op DEPS ${DISTRIBUTE_DEPS})
     set_source_files_properties(recv_op.cc PROPERTIES COMPILE_FLAGS ${DISTRIBUTE_COMPILE_FLAGS})
-    cc_test(test_send_recv SRCS send_recv_op_test.cc DEPS send_op recv_op sum_op executor)
+    op_library(listen_and_serv_op DEPS ${DISTRIBUTE_DEPS})
+    set_source_files_properties(listen_and_serv_op.cc PROPERTIES COMPILE_FLAGS ${DISTRIBUTE_COMPILE_FLAGS})
+    cc_test(test_send_recv SRCS send_recv_op_test.cc DEPS send_op listen_and_serv_op sum_op executor)
 else()
-    set(DEPS_OPS ${DEPS_OPS} send_op recv_op)
+    set(DEPS_OPS ${DEPS_OPS} send_op recv_op listen_and_serv_op)
```
Comment (helinwang) [inline paddle/operators/CMakeLists.txt]: Is `listen_and_serv_op` defined in the case `! WITH_DISTRIBUTE`?

File: paddle/operators/recv_op.cc [hunk 2]
```diff
@@ -12,187 +12,67 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
+ private:
+  mutable detail::RPCClient client_;
 };
 
 class RecvOpMaker : public framework::OpProtoAndCheckerMaker {
  public:
-  RecvOpMaker(OpProto *proto, OpAttrChecker *op_checker)
+  RecvOpMaker(OpProto* proto, OpAttrChecker* op_checker)
       : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "(Tensor) Input tensor to be sent").AsDuplicable();
```
Comment (helinwang) [inline paddle/operators/recv_op.cc]: "X: Input tensor to be sent" is a little confusing, do you mean `X` is the Vars that the send OP will fetch? Btw, `X` is not used in `Run`. Do we need this Input?

File: paddle/operators/recv_op.cc [hunk 3]
```diff
@@ -12,187 +12,67 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-                         "(string, default 127.0.0.1:6164)"
-                         "IP address to listen on.")
-        .SetDefault("127.0.0.1:6164")
-        .AddCustomChecker([](const std::string &ip) { return !ip.empty(); });
-    AddAttr<framework::BlockDesc *>(
-        kOptimizeBlock, "Serialized ProgramDesc string for recv to run.");
-    AddAttr<std::vector<std::string>>(
-        "ParamList", "type list of string",
-        "grad->param name mapping to find which parameters to optimize.")
+    AddAttr<std::vector<std::string>>("endpoints",
```
Comment (helinwang) [inline paddle/operators/recv_op.cc]: Do we need this attribute anymore? It seems not used.

File: paddle/operators/send_op.cc [hunk 4]
```diff
@@ -37,6 +37,7 @@ class SendOp : public framework::OperatorBase {
     auto ins = Inputs("X");
     auto outs = Outputs("Out");
     std::vector<std::string> epmap = Attr<std::vector<std::string>>("epmap");
+    bool do_get = Attr<bool>("DoGet");
     std::vector<std::string> endpoints =
```
Comment (helinwang) [inline paddle/operators/send_op.cc]: `endpoints` seems duplicate with `epmap`, can we remove it. And generate the `endpoints` vector during the iteration of `epmap`?

File: python/paddle/v2/fluid/tests/test_recv_op.py [hunk 5]
```diff
@@ -28,6 +29,7 @@ def test_send(self):
         p = Process(target=self.init_serv, args=(place, ))
         p.daemon = True
         p.start()
+        time.sleep(5)
```
Comment (helinwang) [inline python/paddle/v2/fluid/tests/test_recv_op.py]: Maybe change 5s to something lower (1s?). 5s per unit test is a lot of time considering how much unit tests we run everyday.

File: paddle/operators/send_op.cc [hunk 6]
```diff
@@ -37,6 +37,7 @@ class SendOp : public framework::OperatorBase {
     auto ins = Inputs("X");
     auto outs = Outputs("Out");
     std::vector<std::string> epmap = Attr<std::vector<std::string>>("epmap");
+    bool do_get = Attr<bool>("DoGet");
```
Comment (helinwang) [inline paddle/operators/send_op.cc]: Can we infer `do_get` from `outs.size() > 0`? So we don't need an additional attribute.

File: paddle/operators/CMakeLists.txt [hunk 7]
```diff
@@ -122,9 +122,11 @@ if(WITH_DISTRIBUTE)
     set_source_files_properties(send_op.cc PROPERTIES COMPILE_FLAGS ${DISTRIBUTE_COMPILE_FLAGS})
     op_library(recv_op DEPS ${DISTRIBUTE_DEPS})
     set_source_files_properties(recv_op.cc PROPERTIES COMPILE_FLAGS ${DISTRIBUTE_COMPILE_FLAGS})
-    cc_test(test_send_recv SRCS send_recv_op_test.cc DEPS send_op recv_op sum_op executor)
+    op_library(listen_and_serv_op DEPS ${DISTRIBUTE_DEPS})
+    set_source_files_properties(listen_and_serv_op.cc PROPERTIES COMPILE_FLAGS ${DISTRIBUTE_COMPILE_FLAGS})
```
Comment (helinwang) [inline paddle/operators/CMakeLists.txt]: Have not see `listen_and_serv_op.cc`, did you forgot to checkin?

**Comment by helinwang** [comment]
I think there is missing file from this PR, Maybe it's better to turn on distributed flag in TeamCity CI, so that we can let CI check for us?

File: paddle/operators/CMakeLists.txt (hunk 7 above)
Comment (gongweibao) [reply paddle/operators/CMakeLists.txt]: I didn't find yesterday also.

File: paddle/operators/CMakeLists.txt (hunk 1 above)
Comment (typhoonzero) [reply paddle/operators/CMakeLists.txt]: No. `listen_and_serv_op` is a new name of `recv_op`. In this PR, `send_op` calls `AsyncSendVariable` RPC call as client side and `recv_op` calls `AsyncGetVariable`.

File: paddle/operators/recv_op.cc (hunk 2 above)
Comment (typhoonzero) [reply paddle/operators/recv_op.cc]: No we don't, will remove.

File: paddle/operators/recv_op.cc (hunk 3 above)
Comment (typhoonzero) [reply paddle/operators/recv_op.cc]: Removed.

File: paddle/operators/send_op.cc (hunk 6 above)
Comment (typhoonzero) [reply paddle/operators/send_op.cc]: That's a great idea! Thanks.

File: paddle/operators/send_op.cc (hunk 4 above)
Comment (typhoonzero) [reply paddle/operators/send_op.cc]: Maybe generating a new vector in each `Run` hurts performance, I'll keep it unchanged in this PR and add an todo comment.

**Comment by typhoonzero** [comment]
@helinwang Thanks for reviewing. Sorry that I missed to add `listen_and_serv_op`.

File: paddle/operators/send_op.cc (hunk 4 above)
Comment (helinwang) [reply paddle/operators/send_op.cc]: Thanks for adding the comment! In terms of the performance, I think we should try avoid premature optimization. If its really a performance issue, we can cache the vector. But I doubt if creating a vector every time will have any performance impact.

File: paddle/operators/listen_and_serv_op.cc [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+  void Run(const framework::Scope &scope,
+           const platform::Place &dev_place) const override {
+    platform::DeviceContextPool &pool = platform::DeviceContextPool::Instance();
+    auto &dev_ctx = *pool.Get(dev_place);
+    framework::Scope &recv_scope = scope.NewScope();
+
+    // FIXME(Yancey1989): initialize rpc server with lazy mode.
+    rpc_service_->SetScope(&recv_scope);
+    rpc_service_->SetDevCtx(&dev_ctx);
+    auto param_list = Attr<std::vector<std::string>>("ParamList");
```
Comment (helinwang) [inline paddle/operators/listen_and_serv_op.cc]: Maybe we need to remove `ParamList` and `GradList` attribute. It's only used for printing log:
```
VLOG(3) << "received grad: " << grad_var_name << " updating param: " << param_var_name;
```
I think `listen and serve OP` should be something more general: accepts any input specified by the send OP, run the predefined block, serves any requested variable by the recv OP. So we probably don't need this gradient related log anymore.

File: paddle/operators/listen_and_serv_op.cc [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+          if (it != grad_list.end()) {
+            param_var_name = param_list[it - grad_list.begin()];
+          } else {
+            LOG(ERROR) << "grad has no paired param:" << grad_var_name;
+          }
+          VLOG(3) << "received grad: " << grad_var_name
+                  << " updating param: " << param_var_name;
+
+          if (fan_in > 1) {
+            grad_var_name = this->GetGradVarNameForTrainer(grad_var_name);
```
Comment (helinwang) [inline paddle/operators/listen_and_serv_op.cc]: Not a blocker for merging this PR:

I think the transpiler should handle this naming logic, rather than baked in the listen and serve operator.
If the trainspiler handles this naming logic, we don't need the fan_in variable anymore too: the operator can just wait all the vars being received, and run the optimize block. In this way the retry could be handled better as well: any retry will just override the previous sent variable.

Comment (gongweibao) [reply paddle/operators/listen_and_serv_op.cc]: I think name the `ProgramDesc` maybe better than name the variable.
- Name the variable need to hold the relationship between original and the expansion.It should be held in trainer and pserver also.

What we want to do is distinguish where the variables come from, so an attribute of send_op to can do this.

And `fan_in` is useful when scaling down.
- The Kubernetes Job is a set, and we don't know which trainer will be killed when we set Job's parallelism.

Comment (helinwang) [reply paddle/operators/listen_and_serv_op.cc]: > The Kubernetes Job is a set, and we don't know which trainer will be killed when we set smaller Job's parallelism.fan_in is a count and doesn't care who it is.

We can achieve scaling down without fan_in as well. If we don't change fan_in, then the trainer will block because of the scale down. If we change fan_in, that means a new ProgramDesc is running on the pserver, we can do whatever we want in the new ProgramDesc, not just limited to changing fan_in only.

> I think name the ProgramDesc maybe better than name the variable.

Sorry what do you mean by "name the ProgramDesc"? And why naming the variable needs more memory?

Comment (typhoonzero) [reply paddle/operators/listen_and_serv_op.cc]: Thanks @helinwang 

> I think the transpiler should handle this naming logic (e.g., the gradient for variable W will be W.trainer.0, and W.trainer.1 on trainer 0 and trainer 1 respectively), rather than baked in the listen and serve operator.

That's exactly what I was thinking how to remove the executor `create_var` argument. I'll create an issue to do that in another PR.
